zkml/docs/src/llms-layers/qkv.md

```diff
@@ -0,0 +1,95 @@
+# QKV
+
+## Description of the Layer
+The **QKV** layer is the first layer found in each encoder/decoder of an LLM architecture, and it is employed to produce, for each word (or better, *token*) being processed by the LLM three vectors:
+
+- A query vector `q`
+- A key vector `k`
+- A value vector `v`
+
+These vectors are all computed from a vector `x` representing a token processed by the LLM. More specifically, each of this vector is obtained by multiplying `x` with corresponding weight matrices  $W_q$, $W_k$ and $W_v$, and then each resulting vector is added to a corresponding bias vector $B_q$, $B_k$ and $B_v$. Since an LLM usually needs to process more than 1 token at the same time, the computation of `q`, `k` and `v` vectors for all the input tokens is done in a "batched" fashon over matrices rather than over single vectors:
+
+- The `s` vectors $x_i \in \mathbb{R}^m$, each representing one of the `s` input tokens being processed, are concatenated in a matrix $X \in \mathbb{R}^{s \times e}$, where the $i$-th row of $X$ is equal to $X_i$
+- Define the unitary vector as $\mathbb{1}_s = 1^s$. Given the bias vectors $B_q \in \mathbb{R}^h$, $B_k \in \mathbb{R}^h$ and $B_v \in \mathbb{R}^h$, define the *bias matrices* $B_q^s \in \mathbb{R}^{s \times h}= \mathbb{1}_s*B_q^T$, $B_k^s \in \mathbb{R}^{s \times h}= \mathbb{1}_s*B_k^T$ and $B_v^s \in \mathbb{R}^{s \times h}= \mathbb{1}_s*B_v^T$; in other words, each of the $s$ rows in a bias matrix is equal to the corresponding bias vector 
```

Comment:
I don't think that's the case. For me:
There is only "one" bias vector for the whole Q K V, not one per seq_len. So you multiply [seq_len, emb_size] input with Q for example of shape [emb_size,hidden_size] so the resulting matrix is [seq_len, hidden_size].
Then it's applying the _same_ bias row-wise on each of the seq_len rows. Bias is of size hidden_size.

Reply:
That's the idea, I just wanted to write it in matrix form. It should be clearer now that I have also added the diagram hopefully

Reply:
> each of the $s$ rows in a bias matrix is equal to the corresponding bias vector 

That's not what this sentence is expressing to me at least. "the corresponding bias vector" suggests there is one per row. Which is not true, it's the same for all rows of the same matrix.